components/server/src/api/workspace-service-api.ts

```diff
@@ -174,4 +185,107 @@ export class WorkspaceServiceAPI implements ServiceImpl<typeof WorkspaceServiceI
         response.workspace = this.apiConverter.toWorkspace(info);
         return response;
     }
+
+    async stopWorkspace(req: StopWorkspaceRequest): Promise<StopWorkspaceResponse> {
+        if (!req.workspaceId) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "workspaceId is required");
+        }
+        await this.workspaceService.stopWorkspace(ctxUserId(), req.workspaceId, "stopped via API");
+        const response = new StopWorkspaceResponse();
+        return response;
+    }
+
+    async deleteWorkspace(req: DeleteWorkspaceRequest): Promise<DeleteWorkspaceResponse> {
+        if (!req.workspaceId) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "workspaceId is required");
+        }
+        await this.workspaceService.deleteWorkspace(ctxUserId(), req.workspaceId, "user");
+        const response = new DeleteWorkspaceResponse();
+        return response;
+    }
+
+    async updateWorkspace(request: UpdateWorkspaceRequest): Promise<UpdateWorkspaceResponse> {
+        if (!request.workspaceId) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "workspaceId is required");
+        }
+        const userId = ctxUserId();
+
+        // check if user can access workspace first
+        await this.workspaceService.getWorkspace(userId, request.workspaceId);
```

Comment:
is it really necessary? it is additional DB lookup, I assume indivdual update methods below already have permission checks

Reply:
It's necessary, if user don't have permission or ws not found, then we can avoid to call updates lots of times then get the same error message.